<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · TrajGWAS.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://OpenMendel.github.io/TrajGWAS.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>TrajGWAS.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Example-data-sets"><span>Example data sets</span></a></li><li><a class="tocitem" href="#Basic-usage"><span>Basic usage</span></a></li><li><a class="tocitem" href="#VCF-Formatted-Files"><span>VCF Formatted Files</span></a></li><li><a class="tocitem" href="#BGEN-Formatted-Files"><span>BGEN Formatted Files</span></a></li><li><a class="tocitem" href="#SNP-models"><span>SNP models</span></a></li><li><a class="tocitem" href="#SNP-and/or-sample-masks"><span>SNP and/or sample masks</span></a></li><li><a class="tocitem" href="#Estimating-Effect-Sizes-(Wald)"><span>Estimating Effect Sizes (Wald)</span></a></li><li><a class="tocitem" href="#Score-test-for-screening,-Wald-for-effect-size-estimates"><span>Score test for screening, Wald for effect size estimates</span></a></li><li><a class="tocitem" href="#GxE-or-other-interactions"><span>GxE or other interactions</span></a></li><li><a class="tocitem" href="#SNP-set-testing"><span>SNP-set testing</span></a></li><li><a class="tocitem" href="#Matching-Indicies"><span>Matching Indicies</span></a></li><li><a class="tocitem" href="#Plotting-Results"><span>Plotting Results</span></a></li><li><a class="tocitem" href="#Multiple-Plink-file-sets"><span>Multiple Plink file sets</span></a></li><li><a class="tocitem" href="#Multiple-file-sets-on-cluster"><span>Multiple file sets on cluster</span></a></li><li><a class="tocitem" href="#Troubleshooting"><span>Troubleshooting</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/TrajGWAS.jl/blob/main/docs/src/index.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="TrajGWAS.jl"><a class="docs-heading-anchor" href="#TrajGWAS.jl">TrajGWAS.jl</a><a id="TrajGWAS.jl-1"></a><a class="docs-heading-anchor-permalink" href="#TrajGWAS.jl" title="Permalink"></a></h1><p>TrajGWAS.jl is a Julia package for performing genome-wide association studies (GWAS) for continuous longitudinal phenotypes using a modified linear mixed effects model. It builds upon the <a href="https://github.com/OpenMendel/WiSER.jl">within-subject variance estimation by robust regression (WiSER) method</a> and can be used to identify variants associated with changes in the mean and within-subject variability of the longitduinal trait. The estimation procedure is robust to distributional misspecifications of both the random effects and the response. A saddlepoint approximation (SPA) option is implemented to provide improved power and calibrated type I error for rare variants. </p><p>TrajGWAS.jl currently supports <a href="https://zzz.bwh.harvard.edu/plink/">PLINK</a>, <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">VCF</a> (both dosage and genotype data), and <a href="https://www.well.ox.ac.uk/~gav/bgen_format/">BGEN</a> file formats. We plan to add <a href="https://www.cog-genomics.org/plink/2.0/formats#pgen">PGEN</a> support in the future. </p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>This package requires Julia v1.6 or later. The package is registered and can be installed using the command:</p><pre><code class="language-julia hljs">using Pkg
pkg&quot;add TrajGWAS&quot;</code></pre><p>To run the code in this document, the packages installed by the following command are also necessary:</p><pre><code class="language-julia hljs">pkg&quot;add BenchmarkTools CSV Glob PrettyTables&quot;</code></pre><pre><code class="language-julia hljs"># machine information for this tutorial
versioninfo()</code></pre><pre><code class="nohighlight hljs">Julia Version 1.7.1
Commit ac5cc99908 (2021-12-22 19:35 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.5.0)
  CPU: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)</code></pre><pre><code class="language-julia hljs"># for use in this tutorial
ENV[&quot;COLUMNS&quot;] = 400
using BenchmarkTools, CSV, Glob, SnpArrays, TrajGWAS, PrettyTables</code></pre><h2 id="Example-data-sets"><a class="docs-heading-anchor" href="#Example-data-sets">Example data sets</a><a id="Example-data-sets-1"></a><a class="docs-heading-anchor-permalink" href="#Example-data-sets" title="Permalink"></a></h2><p>The <code>data</code> folder of the package contains the example data sets for use with PLINK and VCF Files. In general, the user can locate this folder by command:</p><pre><code class="language-julia hljs">pvalpath = &quot;trajgwas.pval.txt&quot;
nullpath = &quot;trajgwas.null.txt&quot;
const datadir = normpath(joinpath(dirname(pathof(TrajGWAS)), &quot;../data/&quot;))</code></pre><pre><code class="nohighlight hljs">&quot;/Users/xyz/.julia/dev/TrajGWAS/data/&quot;</code></pre><pre><code class="language-julia hljs"># content of the data folder
readdir(glob&quot;*.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">14-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/bgen_snpsetfile.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/covariate.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/example.8bits.bgen&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/example.8bits.bgen.bgi&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap_snpsetfile.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/sim_data.jl&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/snpsetfile_vcf.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/test_vcf.vcf.gz&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/trajgwas_bgen_ex.csv&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/trajgwas_plinkex.csv&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/trajgwas_vcfex.csv&quot;</code></pre><p>The <code>hapmap3</code> files <code>trajgwas_plinkex.csv</code> file correspond to data examples using PLINK formatted files (.bed, .bim, .fam). </p><p>The <code>test_vcf.vcf.gz</code> and <code>trajgwas_vcfex.csv</code> files are for an example analysis using VCF formatted files. </p><p>The <code>example.8bits.bgen</code> and <code>trajgwas_bgen_ex.csv</code> files are for an example analysis using VCF formatted files. </p><h2 id="Basic-usage"><a class="docs-heading-anchor" href="#Basic-usage">Basic usage</a><a id="Basic-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-usage" title="Permalink"></a></h2><p>The following command performs GWAS using for the hapmap3 PLINK files. The output is the fitted null model.</p><p>The default type of GWAS performed is a single-snp significance genome-wide scan, this can be changed by the keyword <code>analysistype</code> (default is &quot;singlesnp&quot;). Other types of analyses are covered later. It outputs the null model, runtime of fitting the null model, and convergence metrics. </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = pvalpath,
        nullfile = nullpath)</code></pre><pre><code class="nohighlight hljs">******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.602286
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040254






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><p>For documentation of the <code>trajgwas</code> function, type <code>?trajgwas</code> in Julia REPL.</p><article class="docstring"><header><a class="docstring-binding" id="TrajGWAS.trajgwas" href="#TrajGWAS.trajgwas"><code>TrajGWAS.trajgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">trajgwas(nullmeanformula, reformula, nullwsvarformula, idvar, covfile, geneticfile; kwargs...)
trajgwas(nullmeanformula, reformula, nullwsvarformula, idvar, df, geneticfile; kwargs...)
trajgwas(fittednullmodel, geneticfile; kwargs...)
trajgwas(fittednullmodel, bedfile, bimfile, bedn; kwargs...)
trajgwas(fittednullmodel, vcffile, nsamples, vcftype; kwargs...)
trajgwas(fittednullmodel, bgenfile, nsamples; kwargs...)</code></pre><p><strong>Positional arguments</strong></p><ul><li><code>nullmeanformula::FormulaTerm</code>: mean formula (β) for the null model.</li><li><code>reformula::FormulaTerm</code>: random effects formula (γ) for the model.</li><li><code>nullmeanformula::FormulaTerm</code>: within-subject variance formula (τ) for the null model.</li><li><code>idvar::Union{String, Symbol}</code>: id variable for groupings.</li><li><code>covfile::AbstractString</code>: covariate file (csv) with one header line. One column   should be the longitudinal phenotype.</li><li><code>df::DataFrame</code>: DataFrame containing response and regressors for null model.</li><li><code>geneticfile::Union{Nothing, AbstractString}</code>: File containing genetic information for GWAS.   This includes a PLINK file name without the .bed, .fam, or .bim   extensions or a VCF file without the .vcf extension. If <code>geneticfile==nothing</code>,   only null model is fitted. If <code>geneticfile</code> is provided, bed, bim, and fam file (or vcf) with   the same <code>geneticfile</code> prefix need to exist. Compressed file formats such as gz and bz2   are allowed. Check all allowed formats by <code>SnpArrays.ALLOWED_FORMAT</code>. If you&#39;re using a VCF file,   make sure to use the <code>geneticformat = &quot;VCF&quot;</code> keyword option, and specificy dosage (:DS) or   genotype (:GT) data with the <code>vcftype</code> command.</li><li><code>fittednullmodel::StatsModels.TableRegressionModel</code>: the fitted null model   output from <code>trajgwas(nullformula, covfile)</code> or <code>trajgwas(nullformula, df)</code>.   <strong>NOTE</strong> If the nullmodel is passed in with the <code>bedfile, bimfile, bedn</code> or    <code>vcffile, nsamples, vcftype</code> arguments, the IDs/data in the null model must match the   order of the IDs in the PLINK/VCF file.  </li><li><code>bedfile::Union{AbstractString,IOStream}</code>: path to Plink bed file with full file name.</li><li><code>bimfile::Union{AbstractString,IOStream}</code>: path to Plink bim file with full file name.</li><li><code>bedn::Integer</code>: number of samples in bed/vcf file.</li></ul><p><strong>Keyword arguments</strong></p><ul><li><code>analysistype</code>::AbstractString: Type of analysis to conduct. Default is <code>singlesnp</code>. Other options are <code>snpset</code> and <code>gxe</code>.</li><li><code>geneticformat</code>::AbstractString: Type of file used for the genetic analysis. <code>&quot;PLINK&quot;</code> and <code>&quot;VCF&quot;</code> are currently supported. Default is PLINK.</li><li><code>vcftype</code>::Union{Symbol, Nothing}: Data to extract from the VCF file for the GWAS analysis. <code>:DS</code> for dosage or <code>:GT</code> for genotypes. Default is nothing.</li><li><code>nullfile::Union{AbstractString, IOStream}</code>: output file for the fitted null model;   default is <code>trajgwas.null.txt</code>.</li><li><code>pvalfile::Union{AbstractString, IOStream}</code>: output file for the gwas p-values; default is   <code>trajgwas.pval.txt</code>.</li><li><code>covtype::Vector{DataType}</code>: type information for <code>covfile</code>. This is useful   when <code>CSV.read(covarfile)</code> has parsing errors.</li><li><code>covrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for covariate file.</li><li><code>testformula::FormulaTerm</code>: formula for test unit. Default is <code>@formula(trait ~ 0 + snp)</code>.</li><li><code>test::Symbol</code>: <code>:score</code> (default) or <code>:wald</code>.</li><li><code>link::GLM.Link</code>: <code>LogitLink()</code> (default), <code>ProbitLink()</code>, <code>CauchitLink()</code>,   or <code>CloglogLink()</code>.</li><li><code>snpmodel</code>: <code>ADDITIVE_MODEL</code> (default), <code>DOMINANT_MODEL</code>, or <code>RECESSIVE_MODEL</code>.</li><li><code>snpinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: SNP indices for bed/vcf file.</li><li><code>geneticrowinds::Union{Nothing,AbstractVector{&lt;:Integer}}</code>: sample indices for bed/vcf file.</li><li><code>samplepath::Union{Nothing, AbstractString}</code>: path for BGEN sample file if it&#39;s not encoded in the BGEN file.</li><li><code>solver</code>: an optimizer supported by MathOptInterface. Default is   <code>Ipopt.Optimizer()</code>.</li><li><code>solver_config</code>: Dict of configuration keys and values for the optimizer. Default is   <code>Dict(&quot;print_level&quot; =&gt; 0, &quot;mehrotra_algorithm&quot; =&gt; &quot;yes&quot;, &quot;warm_start_init_point&quot; =&gt; &quot;yes&quot;,   &quot;max_iter&quot; =&gt; 100)</code>.</li><li><code>runs::Integer</code>: Number of weighted NLS runs for the null model; default is 2.   Each run will use the newest <code>m.τ</code> and <code>m.Lγ</code> to update the weight matrices   <code>Vi</code> and solve the new weighted NLS.</li><li><code>parallel::Bool</code>: Multi-threading or not for fitting the null model. Default is <code>false</code>.</li><li><code>verbose::Bool</code>: default is <code>false</code>.</li><li><code>snpset::Union{Nothing, Integer, AbstractString, AbstractVector{&lt;:Integer}}</code>: Only include   if you are conducting a snpset analysis. An integer indicates a window of SNPs   (i.e. every 500 snps). An abstract string allows you to specify an input file,   with no header and two columns separated by a space. The first column must contain the snpset ID   and the second column must contain the snpid&#39;s identical to the bimfile. An AbstractVector   allows you to specify the snps you want to perform one joint snpset test for.</li><li><code>e::Union{AbstractString,Symbol}</code>: Only include if you are conducting a GxE analysis.   Enviromental variable to be used to test the GxE interaction.   For instance, for testing <code>sex &amp; snp</code> interaction, use <code>:sex</code> or <code>&quot;sex&quot;</code>.</li></ul><p><strong>Examples</strong></p><p>The following is an example of basic GWAS with PLINK files:</p><pre><code class="language-julia hljs">plinkfile = &quot;plinkexample&quot;
covfile = &quot;covexample&quot;
trajgwas(@formula(trait ~ sex), covfile, plkfile)</code></pre><p>The following is an example of basic GWAS with a VCF file using dosages then genotypes:</p><pre><code class="language-julia hljs">vcffile = &quot;vcfexample&quot;
covfile = &quot;covexample&quot;
trajgwas(@formula(trait ~ sex), covfile, vcffile;
    geneticfile = &quot;VCF&quot;, vcftype = :DS)

trajgwas(@formula(trait ~ sex), covfile, vcffile;
    geneticfile = &quot;VCF&quot;, vcftype = :GT)</code></pre><p>The following is an example of snpset GWAS (every 50 snps). For more types of snpset analyses see documentation:</p><pre><code class="language-julia hljs">trajgwas(@formula(trait ~ sex), covfile, plkfile;
    analysistype = &quot;snpset&quot;, snpset = 50)</code></pre><p>The following is an example of GxE GWAS testing the interaction effect:</p><pre><code class="language-julia hljs">trajgwas(@formula(trait ~ sex), covfile, plkfile;
    analysistype = &quot;gxe&quot;, e = :sex)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/TrajGWAS.jl/blob/0936807cbfd9b19d9e8fbfd18554a95bac6b9ca0/src/gwas.jl#LL23-L127">source</a></section></article><h3 id="Formula-for-null-model"><a class="docs-heading-anchor" href="#Formula-for-null-model">Formula for null model</a><a id="Formula-for-null-model-1"></a><a class="docs-heading-anchor-permalink" href="#Formula-for-null-model" title="Permalink"></a></h3><p>The first three arguments specify the null model without SNP effects. The fourth argument specifies the grouping variable (subject ID, cluster, etc.).</p><ul><li>The first term is the mean fixed effects formula e.g., <code>@formula(y ~ 1 + sex + onMeds)</code>.</li><li>The second argument is the random (location) effects – <code>@formula(y ~ 1)</code> specifies a random intercept, whereas <code>@formula(y ~ 1 + time)</code> would specify a random intercept and slope (time).</li><li>The third argument specifies the within-subject variance (WSV) formula <code>@formula(y ~ 1 + sex + onMeds)</code> models the intra-individual variability of <code>y</code> with an intercept, sex, and medication. Note: The WSV formula must have an intercept. </li><li>The fourth argument specifies the grouping variable of the longitudinal data. In the example data, the <code>id</code> variable specifies the grouping variable the repeated measures are collected on. </li></ul><h3 id="Input-files"><a class="docs-heading-anchor" href="#Input-files">Input files</a><a id="Input-files-1"></a><a class="docs-heading-anchor-permalink" href="#Input-files" title="Permalink"></a></h3><p><code>trajgwas</code> expects two input files: one for responses plus covariates (fifth argument) in long format, the other the genetic file(s) for dosages/genotypes (sixth argument).</p><h4 id="Covariate-and-trait-file"><a class="docs-heading-anchor" href="#Covariate-and-trait-file">Covariate and trait file</a><a id="Covariate-and-trait-file-1"></a><a class="docs-heading-anchor-permalink" href="#Covariate-and-trait-file" title="Permalink"></a></h4><p>Covariates and phenotype are provided in a csv file, e.g., <code>trajgwas_plinkex.csv</code>, which has one header line for variable names. In this example, variable <code>y</code> is continuous variable with repeated measures on individuals specified by <code>id</code>. We want to include variable <code>sex</code> and <code>onMeds</code> as covariates for GWAS.</p><pre><code class="language-julia hljs">run(`head $(datadir)trajgwas_plinkex.csv`);</code></pre><pre><code class="nohighlight hljs">sex,onMeds,snp1,snp2,snp3,snp4,y,id
0.0,1.0,0.0,1.0,2.0,0.0,12.26667411332518,A1
0.0,0.0,0.0,1.0,2.0,0.0,10.268123812744903,A1
0.0,0.0,0.0,1.0,2.0,0.0,12.165997408822557,A1
0.0,0.0,0.0,1.0,2.0,0.0,11.879709602937222,A1
0.0,0.0,0.0,1.0,2.0,0.0,12.812705165990007,A1
0.0,0.0,0.0,1.0,2.0,0.0,9.987659617201372,A1
0.0,0.0,0.0,1.0,2.0,0.0,12.140779426464974,A1
0.0,1.0,0.0,1.0,2.0,0.0,13.205778146177705,A1
0.0,0.0,0.0,1.0,2.0,0.0,11.363145919060207,A1</code></pre><h4 id="Genetic-file(s)"><a class="docs-heading-anchor" href="#Genetic-file(s)">Genetic file(s)</a><a id="Genetic-file(s)-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-file(s)" title="Permalink"></a></h4><p>TrajGWAS supports PLINK files, VCF files, and BGEN files.</p><ul><li><p>Genotype data is available as binary PLINK files.</p></li><li><p>TrajGWAS can use dosage or genotype data from VCF files. </p></li><li><p>TrajGWAS uses dosage data from BGEN files.</p></li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>By default, TrajGWAS assumes a set of PLINK files will be used. When using a VCF File, VCF file and type of data (dosage, genotype) must be specified by the <code>geneticformat</code> and <code>vcftype</code> options (as shown later). Similarly, BGEN must be specified as the <code>geneticformat</code> if using a BGEN file. </p></div></div><pre><code class="language-julia hljs">readdir(glob&quot;hapmap3.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">3-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.fam&quot;</code></pre><p>In this example, there are 324 samples at 13,928 SNPs.</p><pre><code class="language-julia hljs">size(SnpArray(datadir * &quot;hapmap3.bed&quot;))</code></pre><pre><code class="nohighlight hljs">(324, 13928)</code></pre><p>Compressed PLINK and VCF files are supported. For example, if Plink files are <code>hapmap3.bed.gz</code>, <code>hapmap3.bim.gz</code> and <code>hapmap3.fam.gz</code>, the same command</p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = pvalpath,
        nullfile = nullpath)</code></pre><p>still works. Check all supported compression format by</p><pre><code class="language-julia hljs">SnpArrays.ALLOWED_FORMAT</code></pre><pre><code class="nohighlight hljs">6-element Vector{String}:
 &quot;gz&quot;
 &quot;zlib&quot;
 &quot;zz&quot;
 &quot;xz&quot;
 &quot;zst&quot;
 &quot;bz2&quot;</code></pre><h3 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h3><p><code>trajgwas</code> outputs two files: <code>trajgwas.null.txt</code> and <code>trajgwas.pval.txt</code>. </p><ul><li><code>trajgwas.null.txt</code> lists the estimated null model (without SNPs). </li></ul><pre><code class="language-julia hljs">run(`cat trajgwas.null.txt`);</code></pre><pre><code class="nohighlight hljs">Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><ul><li><code>trajgwas.pval.txt</code> tallies the SNPs, their pvalues, and relevant information on each SNP.<ul><li><code>betapval</code> represents the p-value of the SNP&#39;s effect on the mean of the trait. If <code>spa=true</code> (default), then this is the SPA p-value. If <code>spa=false</code>, then this is the score test beta p-value. </li><li><code>taupval</code> represents the p-value of the SNP&#39;s effect on the within-subject variability of the trait. If <code>spa=true</code> (default), then this is the SPA p-value. If <code>spa=false</code>, then this is the score test tau p-value. </li><li><code>jointpval</code> represents a joint p-value of the SNP&#39;s effect on both the mean and variance. By default <code>spa=true</code> this is the harmonic mean of the saddlepoint approximated p-values for beta and tau. If <code>spa=false</code>, this is the joint score test p-value.</li></ul></li></ul><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;trajgwas.pval.txt&quot;, DataFrame), 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m     pos [0m|[1m      snpid [0m|[1m        maf [0m|[1m     hwepval [0m|[1m    betapval [0m|[1m betadir [0m|[1m     taupval [0m|[1m taudir [0m|[1m   jointpval [0m|
|[90m Int64 [0m|[90m   Int64 [0m|[90m   String31 [0m|[90m    Float64 [0m|[90m     Float64 [0m|[90m     Float64 [0m|[90m   Int64 [0m|[90m     Float64 [0m|[90m  Int64 [0m|[90m     Float64 [0m|
|-------|---------|------------|------------|-------------|-------------|---------|-------------|--------|-------------|
|     1 |  554484 | rs10458597 |        0.0 |         1.0 |         1.0 |       0 |         1.0 |      0 |         1.0 |
|     1 |  758311 | rs12562034 |  0.0776398 |    0.409876 |     0.38939 |      -1 |    0.922822 |     -1 |    0.547682 |
|     1 |  967643 |  rs2710875 |   0.324074 |  4.07625e-7 |  5.83856e-6 |       1 |  3.35408e-6 |     -1 |  4.93598e-7 |
|     1 | 1168108 | rs11260566 |   0.191589 |    0.128568 | 0.000850889 |       1 |   0.0559055 |      1 |  0.00101827 |
|     1 | 1375074 |  rs1312568 |   0.441358 |  2.5376e-19 | 0.000171683 |      -1 | 1.84811e-13 |      1 | 2.07091e-15 |
|     1 | 1588771 | rs35154105 |        0.0 |         1.0 |         1.0 |       0 |         1.0 |      0 |         1.0 |
|     1 | 1789051 | rs16824508 | 0.00462963 |    0.933278 |    0.295035 |       1 |    0.304109 |      1 |    0.299504 |
|     1 | 1990452 |  rs2678939 |   0.453704 | 5.07696e-11 |  1.27871e-7 |       1 |  7.73809e-9 |     -1 |   3.3986e-9 |</code></pre><p>Output file names can be changed by the <code>nullfile</code> and <code>pvalfile</code> keywords respectively. For example, </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;trajgwas.pval.txt.gz&quot;,
        nullfile = nullpath)
</code></pre><p>will output the p-value file in compressed gz format.</p><h3 id="Subsamples"><a class="docs-heading-anchor" href="#Subsamples">Subsamples</a><a id="Subsamples-1"></a><a class="docs-heading-anchor-permalink" href="#Subsamples" title="Permalink"></a></h3><p>Use the keyword <code>covrowinds</code> to specify selected samples in the covarite file. Use the keyword <code>geneticrowinds</code> to specify selected samples in the Plink (.bed), VCF, or BGEN file. For example, to use the first 300 samples in both covariate and bed file:</p><pre><code class="language-julia hljs">trajgwas(@formula(trait ~ 1+ sex), 
    @formula(trait ~ 1), 
    @formula(trait ~ 1 + sex), 
    :id,
    covfile, geneticfile, 
    covrowinds=1:300, geneticrowinds=1:300)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should make sure that the selected samples in covariate file match exactly those in bed file, otherwise an error will be displayed. </p></div></div><h3 id="Input-non-genetic-data-as-DataFrame"><a class="docs-heading-anchor" href="#Input-non-genetic-data-as-DataFrame">Input non-genetic data as DataFrame</a><a id="Input-non-genetic-data-as-DataFrame-1"></a><a class="docs-heading-anchor-permalink" href="#Input-non-genetic-data-as-DataFrame" title="Permalink"></a></h3><p>Internally <code>trajgwas</code> parses the covariate file as a DataFrame by <code>df = CSV.read(covfile, DataFrame)</code>. For covariate file of other formats, or for specifying default levels of categorical variables, users can parse their datafile as a DataFrame and then input the DataFrame to <code>trajgwas</code> directly.</p><pre><code class="language-julia hljs">trajgwas(@formula(trait ~ 1+ sex), 
    @formula(trait ~ 1), 
    @formula(trait ~ 1 + sex),
    :id,
    df, geneticfile)</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Users should always make sure that individuals in covariate file or DataFrame match those in Plink fam file/VCF File/BGEN file. </p></div></div><p>For example, following code checks that the subject ID of the <code>covariate.txt</code> file match that of the <code>hapmap3.fam</code> file exactly.</p><pre><code class="language-julia hljs">covdf = CSV.read(datadir * &quot;covariate.txt&quot;, DataFrame)
plkfam = CSV.read(datadir * &quot;hapmap3.fam&quot;, DataFrame, header=0, delim=&#39; &#39;)
all(covdf[!, 2] .== plkfam[!, 2])</code></pre><pre><code class="nohighlight hljs">true</code></pre><pre><code class="language-julia hljs">pretty_table(first(covdf, 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   famid [0m|[1m   perid [0m|[1m  faid [0m|[1m  moid [0m|[1m   sex [0m|[1m trait [0m|
|[90m String7 [0m|[90m String7 [0m|[90m Int64 [0m|[90m Int64 [0m|[90m Int64 [0m|[90m Int64 [0m|
|---------|---------|-------|-------|-------|-------|
|    2431 | NA19916 |     0 |     0 |     1 |     4 |
|    2424 | NA19835 |     0 |     0 |     2 |     4 |
|    2469 | NA20282 |     0 |     0 |     2 |     4 |
|    2368 | NA19703 |     0 |     0 |     1 |     3 |
|    2425 | NA19901 |     0 |     0 |     2 |     3 |
|    2427 | NA19908 |     0 |     0 |     1 |     4 |
|    2430 | NA19914 |     0 |     0 |     2 |     4 |
|    2470 | NA20287 |     0 |     0 |     2 |     1 |</code></pre><pre><code class="language-julia hljs">pretty_table(first(plkfam, 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m Column1 [0m|[1m Column2 [0m|[1m Column3 [0m|[1m Column4 [0m|[1m Column5 [0m|[1m Column6 [0m|
|[90m String7 [0m|[90m String7 [0m|[90m   Int64 [0m|[90m   Int64 [0m|[90m   Int64 [0m|[90m   Int64 [0m|
|---------|---------|---------|---------|---------|---------|
|      A1 | NA19916 |       0 |       0 |       1 |      -9 |
|       2 | NA19835 |       0 |       0 |       2 |      -9 |
|       3 | NA20282 |       0 |       0 |       2 |      -9 |
|       4 | NA19703 |       0 |       0 |       1 |      -9 |
|       5 | NA19901 |       0 |       0 |       2 |      -9 |
|       6 | NA19908 |       0 |       0 |       1 |      -9 |
|       7 | NA19914 |       0 |       0 |       2 |      -9 |
|       8 | NA20287 |       0 |       0 |       2 |      -9 |</code></pre><h3 id="Timing"><a class="docs-heading-anchor" href="#Timing">Timing</a><a id="Timing-1"></a><a class="docs-heading-anchor-permalink" href="#Timing" title="Permalink"></a></h3><p>For this moderate-sized data set, <code>trajgwas</code> takes around 1 second without applying SPA. With SPA, it takes a little longer. </p><pre><code class="language-julia hljs">@btime(trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = pvalpath,
        nullfile = nullpath, 
        usespa = false));</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044445
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.042232
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043275
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039053
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044923
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039103
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044385
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039346
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043468
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039741
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043577
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040535
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043344
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039667
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043612
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039483
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.045086
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039441
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044173
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040099
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044949
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040097
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044071
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.041057
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044740
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040537
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044379
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.042448
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.045716
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.042732
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043419
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040273
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044672
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040251
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.045095
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.041001
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043335
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039804
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044344
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039873
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043518
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039340
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044033
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039883
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043352
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.044078
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043401
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039271
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044151
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039558
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043364
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039549
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044107
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040570
run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043527
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040298
  459.587 ms (1433770 allocations: 127.29 MiB)</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;trajgwas.pval.txt&quot;, force=true)</code></pre><h2 id="VCF-Formatted-Files"><a class="docs-heading-anchor" href="#VCF-Formatted-Files">VCF Formatted Files</a><a id="VCF-Formatted-Files-1"></a><a class="docs-heading-anchor-permalink" href="#VCF-Formatted-Files" title="Permalink"></a></h2><p>By default, TrajGWAS.jl will assume you are using PLINK files. It also supports VCF (and BGEN) Files. <code>vcf.gz</code> files are supported as well. To use vcf files in any of the analysis options detailed in this documentation, you simply need to add two keyword options to the <code>trajgwas</code> function:</p><ul><li><code>geneticformat</code>: Choices are &quot;VCF&quot; or &quot;PLINK&quot;. If you are using a VCF file, use <code>geneticformat = &quot;VCF&quot;</code>.</li><li><code>vcftype</code>: Choices are :GT (for genotypes) or :DS (for dosages). This tells TrajGWAS which type of data to extract from the VCF file.</li></ul><p>Using a VCF File does not output minor allele frequency or hardy weinberg equillibrium p-values for each SNP tested since they may be dosages. </p><p>The following shows how to run an analysis with a VCF file using the dosage information. </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_vcfex.csv&quot;,
        datadir * &quot;test_vcf&quot;,
        pvalfile = pvalpath,
        geneticformat = &quot;VCF&quot;,
        vcftype = :DS)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.003459, ‖Δτ‖ = 0.421747, ‖ΔL‖ = 0.002492, status = LOCALLY_SOLVED, time(s) = 0.020257
run = 2, ‖Δβ‖ = 0.001369, ‖Δτ‖ = 0.015998, ‖ΔL‖ = 0.003244, status = LOCALLY_SOLVED, time(s) = 0.017106






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 191
Total observations: 1910

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  12.1708     0.18351     66.32    &lt;1e-99
β2: sex          -3.41012    0.253155   -13.47    &lt;1e-40
β3: onMeds        0.485447   0.0701423    6.92    &lt;1e-11
τ1: (Intercept)   0.774299   0.0849867    9.11    &lt;1e-19
τ2: sex          -0.524117   0.106719    -4.91    &lt;1e-06
τ3: onMeds        0.633337   0.0814974    7.77    &lt;1e-14
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  2.86452</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;trajgwas.pval.txt&quot;, DataFrame), 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m      pos [0m|[1m       snpid [0m|[1m  betapval [0m|[1m betadir [0m|[1m    taupval [0m|[1m taudir [0m|[1m  jointpval [0m|
|[90m Int64 [0m|[90m    Int64 [0m|[90m    String15 [0m|[90m   Float64 [0m|[90m   Int64 [0m|[90m    Float64 [0m|[90m  Int64 [0m|[90m    Float64 [0m|
|-------|----------|-------------|-----------|---------|------------|--------|------------|
|    22 | 20000086 | rs138720731 |  0.322684 |      -1 |    0.61956 |     -1 |   0.424353 |
|    22 | 20000146 |  rs73387790 |       1.0 |       0 |        1.0 |      0 |        1.0 |
|    22 | 20000199 | rs183293480 |  0.214207 |       1 |   0.190785 |      1 |   0.201819 |
|    22 | 20000291 | rs185807825 |  0.292231 |       1 |   0.200029 |      1 |   0.237495 |
|    22 | 20000428 |  rs55902548 | 0.0140114 |       1 | 0.00369005 |      1 | 0.00576184 |
|    22 | 20000683 | rs142720028 |       1.0 |       0 |        1.0 |      0 |        1.0 |
|    22 | 20000771 | rs114690707 |  0.536406 |      -1 |   0.145113 |      1 |    0.22843 |
|    22 | 20000793 | rs189842693 |  0.161414 |      -1 |   0.757017 |      1 |   0.266091 |</code></pre><h2 id="BGEN-Formatted-Files"><a class="docs-heading-anchor" href="#BGEN-Formatted-Files">BGEN Formatted Files</a><a id="BGEN-Formatted-Files-1"></a><a class="docs-heading-anchor-permalink" href="#BGEN-Formatted-Files" title="Permalink"></a></h2><p>By default, TrajGWAS.jl will assume you are using PLINK files. It also supports BGEN Files. To use BGEN files in any of the analysis options detailed in this documentation, you simply need to add the following keyword option to the <code>trajgwas</code> function:</p><ul><li><code>geneticformat</code>: Choices are &quot;VCF&quot; or &quot;PLINK&quot; or &quot;BGEN&quot;. If you are using a BGEN file, use <code>geneticformat = &quot;BGEN&quot;</code>.</li></ul><p>Using a BGEN File does not output minor allele frequency or hardy weinberg equillibrium p-values for each SNP tested.</p><p>Some features, such as SNP-set analyses, are only available when there&#39;s an indexing <code>.bgi</code> file available. Additionally, if the BGEN file does not contain sample information (i.e. sample IDs), then a sample file is necessary and its path can be specified with the <code>samplepath</code> keyword. </p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>BGEN files can contain an optional index file (<code>.bgi</code> file) that allows the variants to be specified in order of position. TrajGWAS will automatically look for a file in the same directory as the <code>BGENFILENAME</code> with the name <code>BGENFILENAME.bgi</code>. The BGEN file is read either in the <code>.bgi</code> file order if <code>BGENFILENAME.bgi</code> is supplied in the same directory as the BGEN file, otherwise it will use the order in the BGEN file. This is important in analyses specifying <code>snpinds</code> as well as annotation groupings. You must make sure this matches the way the BGEN file will be read in. </p></div></div><p>The following shows how to run an analysis with a BGEN file using the dosage information. </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id, 
        datadir * &quot;trajgwas_bgen_ex.csv&quot;, 
        datadir * &quot;example.8bits&quot;, 
        geneticformat = &quot;BGEN&quot;, 
        pvalfile = pvalpath)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.096578, ‖Δτ‖ = 0.162563, ‖ΔL‖ = 0.007625, status = LOCALLY_SOLVED, time(s) = 0.055072
run = 2, ‖Δβ‖ = 0.003173, ‖Δτ‖ = 0.005584, ‖ΔL‖ = 0.001468, status = LOCALLY_SOLVED, time(s) = 0.040037






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 500
Total observations: 5000

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  10.5275     0.0992675  106.05    &lt;1e-99
β2: sex          -3.38067    0.146416   -23.09    &lt;1e-99
β3: onMeds        0.522162   0.0358727   14.56    &lt;1e-47
τ1: (Intercept)   0.294364   0.0443191    6.64    &lt;1e-10
τ2: sex          -0.365009   0.0503589   -7.25    &lt;1e-12
τ3: onMeds        0.559769   0.0467046   11.99    &lt;1e-32
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  2.51561</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;trajgwas.pval.txt&quot;, DataFrame), 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m   pos [0m|[1m    snpid [0m|[1m     varid [0m|[1m  hwepval [0m|[1m      maf [0m|[1m infoscore [0m|[1m    betapval [0m|[1m betadir [0m|[1m     taupval [0m|[1m taudir [0m|[1m   jointpval [0m|
|[90m Int64 [0m|[90m Int64 [0m|[90m String15 [0m|[90m  String15 [0m|[90m  Float64 [0m|[90m  Float64 [0m|[90m   Float64 [0m|[90m     Float64 [0m|[90m   Int64 [0m|[90m     Float64 [0m|[90m  Int64 [0m|[90m     Float64 [0m|
|-------|-------|----------|-----------|----------|----------|-----------|-------------|---------|-------------|--------|-------------|
|     1 |  1001 | RSID_101 | SNPID_101 |  0.86274 | 0.416977 |  0.984639 |    0.510629 |      -1 |    0.425112 |     -1 |    0.463963 |
|     1 |  2000 |   RSID_2 |   SNPID_2 | 0.192181 |  0.19751 |       9.0 |  4.5062e-10 |       1 | 1.40173e-21 |     -1 | 6.67473e-21 |
|     1 |  2001 | RSID_102 | SNPID_102 |   0.1844 | 0.197667 |  0.727308 | 4.47847e-10 |      -1 | 1.79121e-21 |      1 | 8.45783e-21 |
|     1 |  3000 |   RSID_3 |   SNPID_3 | 0.965354 | 0.483396 |  0.955355 |   0.0721822 |      -1 |    0.600949 |     -1 |    0.128884 |
|     1 |  3001 | RSID_103 | SNPID_103 | 0.965354 | 0.483396 |  0.955355 |   0.0721821 |       1 |    0.600949 |      1 |    0.128884 |
|     1 |  4000 |   RSID_4 |   SNPID_4 | 0.371927 |  0.21671 |  0.991768 |    0.112472 |      -1 |    0.133752 |     -1 |    0.122192 |
|     1 |  4001 | RSID_104 | SNPID_104 | 0.371928 |  0.21671 |  0.991768 |    0.112471 |       1 |    0.133752 |      1 |    0.122192 |
|     1 |  5000 |   RSID_5 |   SNPID_5 | 0.587013 | 0.388082 |  0.968258 |    0.526325 |      -1 |    0.542332 |     -1 |    0.534209 |</code></pre><h2 id="SNP-models"><a class="docs-heading-anchor" href="#SNP-models">SNP models</a><a id="SNP-models-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-models" title="Permalink"></a></h2><p>Genotypes are translated into numeric values according to different genetic model, which is specified by the <code>snpmodel</code> keyword. Default is <code>ADDITIVE_MODEL</code>.</p><table><tr><th style="text-align: center">Genotype</th><th style="text-align: center"><code>SnpArray</code></th><th style="text-align: center"><code>ADDITIVE_MODEL</code></th><th style="text-align: center"><code>DOMINANT_MODEL</code></th><th style="text-align: center"><code>RECESSIVE_MODEL</code></th></tr><tr><td style="text-align: center">A1,A1</td><td style="text-align: center">0x00</td><td style="text-align: center">0</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">missing</td><td style="text-align: center">0x01</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td><td style="text-align: center">NaN</td></tr><tr><td style="text-align: center">A1,A2</td><td style="text-align: center">0x02</td><td style="text-align: center">1</td><td style="text-align: center">1</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">A2,A2</td><td style="text-align: center">0x03</td><td style="text-align: center">2</td><td style="text-align: center">1</td><td style="text-align: center">1</td></tr></table><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>trajgwas</code> imputes missing genotypes according to minor allele frequencies. </p></div></div><p>Users are advised to impute genotypes using more sophiscated methods before GWAS.</p><h2 id="SNP-and/or-sample-masks"><a class="docs-heading-anchor" href="#SNP-and/or-sample-masks">SNP and/or sample masks</a><a id="SNP-and/or-sample-masks-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-and/or-sample-masks" title="Permalink"></a></h2><p>In practice, we often perform GWAS on selected SNPs and/or selected samples. They can be specified by the <code>snpinds</code>, <code>covrowinds</code> and <code>geneticrowinds</code> keywords of <code>trajgwas</code> function. </p><p>For example, to perform GWAS on SNPs with minor allele frequency (MAF) above 0.05</p><pre><code class="language-julia hljs"># create SNP mask
snpinds = maf(SnpArray(&quot;../data/hapmap3.bed&quot;)) .≥ 0.05
# GWAS on selected SNPs
@time trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;commonvariant.pval.txt&quot;,
        snpinds = snpinds)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043349
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039752
  2.733297 seconds (8.07 M allocations: 492.816 MiB, 5.49% gc time, 65.58% compilation time)






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">run(`head commonvariant.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr	pos	snpid	maf	hwepval	betapval	betadir	taupval	taudir	jointpval
1	758311	rs12562034	0.07763975155279501	0.4098763332666681	0.3893898715952268	-1	0.9228220976232073	-1	0.5476822137398478
1	967643	rs2710875	0.32407407407407407	4.076249100705747e-7	5.838562452864651e-6	1	3.354075049151666e-6	-1	4.935983727867445e-7
1	1168108	rs11260566	0.19158878504672894	0.1285682279446898	0.0008508894259509042	1	0.05590549977468929	1	0.0010182674469333282
1	1375074	rs1312568	0.441358024691358	2.5376019650614977e-19	0.00017168324418138014	-1	1.8481146312336014e-13	1	2.0709090755361924e-15
1	1990452	rs2678939	0.4537037037037037	5.07695957708431e-11	1.278709529589261e-7	1	7.7380891475822e-9	-1	3.3986012947082272e-9
1	2194615	rs7553178	0.22685185185185186	0.17056143157457776	0.07404001756940115	-1	0.00023109048716816823	1	0.00010021413726889922
1	2396747	rs13376356	0.1448598130841121	0.9053079215078139	0.48972088680069503	-1	0.6595593862223496	1	0.5620909278620404
1	2823603	rs1563468	0.4830246913580247	4.23065537243926e-9	2.581476982676523e-7	-1	1.1700736777037956e-8	1	6.36797979359385e-8
1	3025087	rs6690373	0.2538699690402477	9.238641887192776e-8	3.111965106232366e-5	1	2.0479935907554532e-7	-1	2.123094937745341e-5</code></pre><pre><code class="language-julia hljs"># extra header line in commonvariant.pval.txt
countlines(&quot;commonvariant.pval.txt&quot;), count(snpinds)</code></pre><pre><code class="nohighlight hljs">(12086, 12085)</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;commonvariant.pval.txt&quot;, force=true)</code></pre><p><code>covrowinds</code> specify the samples in the covariate file and <code>geneticrowinds</code> for PLINK or VCF File. User should be particularly careful when using these two keyword. Selected rows in SnpArray should exactly match the samples in the null model. Otherwise the results are meaningless.</p><h2 id="Estimating-Effect-Sizes-(Wald)"><a class="docs-heading-anchor" href="#Estimating-Effect-Sizes-(Wald)">Estimating Effect Sizes (Wald)</a><a id="Estimating-Effect-Sizes-(Wald)-1"></a><a class="docs-heading-anchor-permalink" href="#Estimating-Effect-Sizes-(Wald)" title="Permalink"></a></h2><p>By default, <code>trajgwas</code> calculates p-value for each SNP using SPA/score test. Score test is fast because it doesn&#39;t require fitting alternative model for each SNP. User can request the Wald p-values and the estimated effect size of each SNP using keyword <code>test=:wald</code>. Wald is much slower but will give you estimated effect sizes from the WiSER model.</p><pre><code class="language-julia hljs">@time trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;wald.pval.txt&quot;,
        snpinds = 1:5,
        test = :wald)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.051852
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.042647
run = 1, ‖Δβ‖ = 0.033243, ‖Δτ‖ = 0.146511, ‖ΔL‖ = 0.005476, status = LOCALLY_SOLVED, time(s) = 0.052239
run = 2, ‖Δβ‖ = 0.005774, ‖Δτ‖ = 0.042246, ‖ΔL‖ = 0.001784, status = LOCALLY_SOLVED, time(s) = 0.048045
run = 1, ‖Δβ‖ = 0.013090, ‖Δτ‖ = 0.130781, ‖ΔL‖ = 0.005011, status = LOCALLY_SOLVED, time(s) = 0.057742
run = 2, ‖Δβ‖ = 0.003913, ‖Δτ‖ = 0.037309, ‖ΔL‖ = 0.001516, status = LOCALLY_SOLVED, time(s) = 0.052004
run = 1, ‖Δβ‖ = 0.022159, ‖Δτ‖ = 0.141135, ‖ΔL‖ = 0.005554, status = LOCALLY_SOLVED, time(s) = 0.048711
run = 2, ‖Δβ‖ = 0.001482, ‖Δτ‖ = 0.021700, ‖ΔL‖ = 0.001435, status = LOCALLY_SOLVED, time(s) = 0.044404
run = 1, ‖Δβ‖ = 0.026764, ‖Δτ‖ = 0.368620, ‖ΔL‖ = 0.000317, status = LOCALLY_SOLVED, time(s) = 0.047687
run = 2, ‖Δβ‖ = 0.003023, ‖Δτ‖ = 0.030938, ‖ΔL‖ = 0.003568, status = LOCALLY_SOLVED, time(s) = 0.043295
  4.293410 seconds (7.17 M allocations: 431.003 MiB, 2.57% gc time, 86.00% compilation time)






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><p>Note the extra <code>effect</code> column in pvalfile, which is the effect size (regression coefficient) for each SNP. </p><pre><code class="language-julia hljs">run(`head wald.pval.txt`);</code></pre><pre><code class="nohighlight hljs">chr	pos	snpid	maf	hwepval	betaeffect	betapval	taueffect	taupval
1	554484	rs10458597	0.0	1.0	0.0	1.0	0.0	1.0
1	758311	rs12562034	0.07763975155279501	0.4098763332666681	-0.231072479332816	0.3929844692486013	0.01340575997086949	0.9216488777690096
1	967643	rs2710875	0.32407407407407407	4.076249100705747e-7	0.6406954913590558	8.824209286611495e-7	-0.34763283794106525	1.0836367479719384e-11
1	1168108	rs11260566	0.19158878504672894	0.1285682279446898	0.6188250819875118	0.0002537468407208455	-0.14748437551458396	0.0502802758129169
1	1375074	rs1312568	0.441358024691358	2.5376019650614977e-19	-0.4560323801786057	0.00016592110001810087	0.4964521486169327	1.177346561221082e-36</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;wald.pval.txt&quot;, force=true)
rm(&quot;trajgwas.null.txt&quot;, force=true)</code></pre><h2 id="Score-test-for-screening,-Wald-for-effect-size-estimates"><a class="docs-heading-anchor" href="#Score-test-for-screening,-Wald-for-effect-size-estimates">Score test for screening, Wald for effect size estimates</a><a id="Score-test-for-screening,-Wald-for-effect-size-estimates-1"></a><a class="docs-heading-anchor-permalink" href="#Score-test-for-screening,-Wald-for-effect-size-estimates" title="Permalink"></a></h2><p>For large data sets, a practical solution is to perform the score test first across all SNPs, then re-do Wald for the most promising SNPs according to score test p-values in order to get estimated effect sizes.</p><p><strong>Step 1</strong>: Perform score test GWAS, results in <code>score.pval.txt</code>.</p><pre><code class="language-julia hljs">@time trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;score.pval.txt&quot;)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044256
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.041231
  1.121874 seconds (4.00 M allocations: 259.048 MiB, 13.18% gc time)






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;score.pval.txt&quot;, DataFrame), 8), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m     pos [0m|[1m      snpid [0m|[1m        maf [0m|[1m     hwepval [0m|[1m    betapval [0m|[1m betadir [0m|[1m     taupval [0m|[1m taudir [0m|[1m   jointpval [0m|
|[90m Int64 [0m|[90m   Int64 [0m|[90m   String31 [0m|[90m    Float64 [0m|[90m     Float64 [0m|[90m     Float64 [0m|[90m   Int64 [0m|[90m     Float64 [0m|[90m  Int64 [0m|[90m     Float64 [0m|
|-------|---------|------------|------------|-------------|-------------|---------|-------------|--------|-------------|
|     1 |  554484 | rs10458597 |        0.0 |         1.0 |         1.0 |       0 |         1.0 |      0 |         1.0 |
|     1 |  758311 | rs12562034 |  0.0776398 |    0.409876 |     0.38939 |      -1 |    0.922822 |     -1 |    0.547682 |
|     1 |  967643 |  rs2710875 |   0.324074 |  4.07625e-7 |  5.83856e-6 |       1 |  3.35408e-6 |     -1 |  4.93598e-7 |
|     1 | 1168108 | rs11260566 |   0.191589 |    0.128568 | 0.000850889 |       1 |   0.0559055 |      1 |  0.00101827 |
|     1 | 1375074 |  rs1312568 |   0.441358 |  2.5376e-19 | 0.000171683 |      -1 | 1.84811e-13 |      1 | 2.07091e-15 |
|     1 | 1588771 | rs35154105 |        0.0 |         1.0 |         1.0 |       0 |         1.0 |      0 |         1.0 |
|     1 | 1789051 | rs16824508 | 0.00462963 |    0.933278 |    0.295035 |       1 |    0.304109 |      1 |    0.299504 |
|     1 | 1990452 |  rs2678939 |   0.453704 | 5.07696e-11 |  1.27871e-7 |       1 |  7.73809e-9 |     -1 |   3.3986e-9 |</code></pre><p><strong>Step 2</strong>: Sort score test p-values and find top 10 SNPs.</p><pre><code class="language-julia hljs">scorepvals = CSV.read(&quot;score.pval.txt&quot;, DataFrame)[!, :taupval] # tau p-values 
tophits = sortperm(scorepvals)[1:10] # indices of 10 SNPs with smallest p-values
scorepvals[tophits] # smallest 10 p-values</code></pre><pre><code class="nohighlight hljs">10-element Vector{Float64}:
 7.872202557706978e-17
 1.8481146312336014e-13
 1.497868141369434e-12
 9.730062616490452e-12
 1.2892059340654512e-11
 1.4844638660225815e-11
 1.5167114834215392e-11
 1.8761256218261456e-11
 2.3491352960094672e-11
 2.6075629334507108e-11</code></pre><p><strong>Step 3</strong>: Re-do LRT on top hits.</p><pre><code class="language-julia hljs">@time trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;wald.pval.txt&quot;,
        snpinds = tophits,
        test = :wald)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043852
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.038897
run = 1, ‖Δβ‖ = 0.026764, ‖Δτ‖ = 0.368620, ‖ΔL‖ = 0.000317, status = LOCALLY_SOLVED, time(s) = 0.051141
run = 2, ‖Δβ‖ = 0.003023, ‖Δτ‖ = 0.030938, ‖ΔL‖ = 0.003568, status = LOCALLY_SOLVED, time(s) = 0.045226
run = 1, ‖Δβ‖ = 0.039946, ‖Δτ‖ = 0.185049, ‖ΔL‖ = 0.005690, status = LOCALLY_SOLVED, time(s) = 0.047814
run = 2, ‖Δβ‖ = 0.001640, ‖Δτ‖ = 0.029639, ‖ΔL‖ = 0.002028, status = LOCALLY_SOLVED, time(s) = 0.047543
run = 1, ‖Δβ‖ = 0.040207, ‖Δτ‖ = 0.407863, ‖ΔL‖ = 0.001924, status = LOCALLY_SOLVED, time(s) = 0.049603
run = 2, ‖Δβ‖ = 0.002632, ‖Δτ‖ = 0.039609, ‖ΔL‖ = 0.005131, status = LOCALLY_SOLVED, time(s) = 0.049566
run = 1, ‖Δβ‖ = 0.052372, ‖Δτ‖ = 0.828946, ‖ΔL‖ = 0.001277, status = LOCALLY_SOLVED, time(s) = 0.061910
run = 2, ‖Δβ‖ = 0.013694, ‖Δτ‖ = 0.185493, ‖ΔL‖ = 0.007118, status = LOCALLY_SOLVED, time(s) = 0.052748
run = 1, ‖Δβ‖ = 0.032435, ‖Δτ‖ = 0.199659, ‖ΔL‖ = 0.004973, status = LOCALLY_SOLVED, time(s) = 0.049459
run = 2, ‖Δβ‖ = 0.002847, ‖Δτ‖ = 0.032438, ‖ΔL‖ = 0.002941, status = LOCALLY_SOLVED, time(s) = 0.048901
run = 1, ‖Δβ‖ = 0.009362, ‖Δτ‖ = 0.625969, ‖ΔL‖ = 0.002237, status = LOCALLY_SOLVED, time(s) = 0.061187
run = 2, ‖Δβ‖ = 0.005893, ‖Δτ‖ = 0.176417, ‖ΔL‖ = 0.005017, status = LOCALLY_SOLVED, time(s) = 0.052250
run = 1, ‖Δβ‖ = 0.012805, ‖Δτ‖ = 0.243313, ‖ΔL‖ = 0.006289, status = LOCALLY_SOLVED, time(s) = 0.051522
run = 2, ‖Δβ‖ = 0.002636, ‖Δτ‖ = 0.044210, ‖ΔL‖ = 0.003312, status = LOCALLY_SOLVED, time(s) = 0.053374
run = 1, ‖Δβ‖ = 0.014062, ‖Δτ‖ = 0.225841, ‖ΔL‖ = 0.003899, status = LOCALLY_SOLVED, time(s) = 0.054011
run = 2, ‖Δβ‖ = 0.002112, ‖Δτ‖ = 0.030955, ‖ΔL‖ = 0.003860, status = LOCALLY_SOLVED, time(s) = 0.053688
run = 1, ‖Δβ‖ = 0.026425, ‖Δτ‖ = 0.258805, ‖ΔL‖ = 0.004204, status = LOCALLY_SOLVED, time(s) = 0.051707
run = 2, ‖Δβ‖ = 0.001090, ‖Δτ‖ = 0.051281, ‖ΔL‖ = 0.003822, status = LOCALLY_SOLVED, time(s) = 0.054497
run = 1, ‖Δβ‖ = 0.035737, ‖Δτ‖ = 0.174702, ‖ΔL‖ = 0.001439, status = LOCALLY_SOLVED, time(s) = 0.053476
run = 2, ‖Δβ‖ = 0.004776, ‖Δτ‖ = 0.019790, ‖ΔL‖ = 0.001867, status = LOCALLY_SOLVED, time(s) = 0.051735
  3.154140 seconds (5.48 M allocations: 365.218 MiB, 3.34% gc time, 58.54% compilation time)






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(CSV.read(&quot;wald.pval.txt&quot;, DataFrame), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m       pos [0m|[1m     snpid [0m|[1m      maf [0m|[1m     hwepval [0m|[1m betaeffect [0m|[1m    betapval [0m|[1m taueffect [0m|[1m     taupval [0m|
|[90m Int64 [0m|[90m     Int64 [0m|[90m  String15 [0m|[90m  Float64 [0m|[90m     Float64 [0m|[90m    Float64 [0m|[90m     Float64 [0m|[90m   Float64 [0m|[90m     Float64 [0m|
|-------|-----------|-----------|----------|-------------|------------|-------------|-----------|-------------|
|     1 |   1375074 | rs1312568 | 0.441358 |  2.5376e-19 |  -0.456032 | 0.000165921 |  0.496452 | 1.17735e-36 |
|     1 |  11552817 | rs2745282 | 0.421053 |  1.3537e-14 |   -0.67282 |   7.6492e-9 |  0.425653 | 1.45395e-19 |
|     1 | 120276030 | rs6688004 | 0.469136 | 1.75473e-27 |  -0.637817 |  5.13062e-9 |  0.430562 | 6.07457e-21 |
|     2 | 135623558 | rs6730157 | 0.313272 | 7.76753e-16 |  -0.556166 |  1.10163e-5 |  0.462346 | 2.88628e-20 |
|     9 | 126307510 | rs3814134 | 0.427469 | 4.28276e-33 |   0.723472 | 1.41601e-11 | -0.518216 | 4.75338e-39 |
|    15 |  40603025 | rs2617236 | 0.390966 | 1.16466e-14 |  -0.644375 |  1.44948e-7 |  0.445647 | 9.43451e-19 |
|    15 |  40803767 | rs3742988 |  0.42284 | 4.26181e-30 |  -0.796743 | 2.24785e-13 |  0.389234 | 7.09349e-17 |
|    17 |  56509992 | rs8064681 | 0.481481 | 1.37485e-21 |   -0.58961 |   1.9256e-7 |  0.425417 | 5.60171e-25 |
|    17 |  71293786 | rs2125345 | 0.339009 | 1.75527e-14 |  -0.635669 |  1.82739e-7 |  0.449612 |  1.7146e-21 |
|    23 |  64815688 | rs5964999 | 0.475078 | 3.63212e-56 |    0.75886 | 5.04109e-14 |  -0.38849 | 5.39098e-23 |</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;trajgwas.pval.txt&quot;, force=true)
rm(&quot;score.pval.txt&quot;, force=true)
rm(&quot;wald.pval.txt&quot;, force=true)</code></pre><h2 id="GxE-or-other-interactions"><a class="docs-heading-anchor" href="#GxE-or-other-interactions">GxE or other interactions</a><a id="GxE-or-other-interactions-1"></a><a class="docs-heading-anchor-permalink" href="#GxE-or-other-interactions" title="Permalink"></a></h2><h3 id="Testing-jointly-G-GxE"><a class="docs-heading-anchor" href="#Testing-jointly-G-GxE">Testing jointly G + GxE</a><a id="Testing-jointly-G-GxE-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-jointly-G-GxE" title="Permalink"></a></h3><p>In many applications, we want to test SNP effect and/or its interaction with other terms. <code>testformula</code> keyword specifies the test unit <strong>besides</strong> the covariates in <code>nullformula</code>. </p><p>In following example, keyword <code>testformula=@formula(trait ~ snp + snp &amp; sex)</code> instructs <code>trajgwas</code> to test joint effect of <code>snp</code> and <code>snp &amp; sex</code> interaction.</p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;GxE.pval.txt&quot;,
        testformula=@formula(trait ~ snp + snp &amp; sex))</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.046905
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.043433






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;GxE.pval.txt&quot;, DataFrame), 5), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m     pos [0m|[1m      snpid [0m|[1m       maf [0m|[1m    hwepval [0m|[1m    betapval [0m|[1m     taupval [0m|[1m   jointpval [0m|
|[90m Int64 [0m|[90m   Int64 [0m|[90m   String31 [0m|[90m   Float64 [0m|[90m    Float64 [0m|[90m     Float64 [0m|[90m     Float64 [0m|[90m     Float64 [0m|
|-------|---------|------------|-----------|------------|-------------|-------------|-------------|
|     1 |  554484 | rs10458597 |       0.0 |        1.0 |         1.0 |         1.0 |         1.0 |
|     1 |  758311 | rs12562034 | 0.0776398 |   0.409876 |       0.383 |     0.33511 |    0.357458 |
|     1 |  967643 |  rs2710875 |  0.324074 | 4.07625e-7 |  9.98217e-6 |  5.73202e-7 |  1.08415e-6 |
|     1 | 1168108 | rs11260566 |  0.191589 |   0.128568 |  0.00225789 |    0.105482 |  0.00442114 |
|     1 | 1375074 |  rs1312568 |  0.441358 | 2.5376e-19 | 0.000928728 | 3.48807e-15 | 6.97615e-15 |</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;GxE.pval.txt&quot;,  force=true)</code></pre><h3 id="Testing-only-GxE-interaction-term"><a class="docs-heading-anchor" href="#Testing-only-GxE-interaction-term">Testing only GxE interaction term</a><a id="Testing-only-GxE-interaction-term-1"></a><a class="docs-heading-anchor-permalink" href="#Testing-only-GxE-interaction-term" title="Permalink"></a></h3><p>For some applications, the user may want to simply test the GxE interaction effect. This requires fitting the SNP in the null model and is much slower, but the command <code>trajgwas()</code> with keyword <code>analysistype = &quot;gxe&quot;</code> can be used test the interaction effect. The environmental variable must be specified in the command using the keyword argument <code>e</code>, either as a symbol, such as <code>:age</code> or as a string <code>&quot;age&quot;</code>. </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;gxe_snp.pval.txt&quot;, 
        analysistype = &quot;gxe&quot;,
        e = :sex, 
        snpinds=1:5)
</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.045341
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040165
run = 1, ‖Δβ‖ = 0.480377, ‖Δτ‖ = 0.023510, ‖ΔL‖ = 0.001822, status = LOCALLY_SOLVED, time(s) = 0.045323
run = 2, ‖Δβ‖ = 0.000323, ‖Δτ‖ = 0.002136, ‖ΔL‖ = 0.000025, status = LOCALLY_SOLVED, time(s) = 0.044147
run = 1, ‖Δβ‖ = 1.059989, ‖Δτ‖ = 0.458239, ‖ΔL‖ = 0.063093, status = LOCALLY_SOLVED, time(s) = 0.060090
run = 2, ‖Δβ‖ = 0.009118, ‖Δτ‖ = 0.105352, ‖ΔL‖ = 0.000056, status = LOCALLY_SOLVED, time(s) = 0.056690
run = 1, ‖Δβ‖ = 1.144781, ‖Δτ‖ = 0.251459, ‖ΔL‖ = 0.034318, status = LOCALLY_SOLVED, time(s) = 0.047379
run = 2, ‖Δβ‖ = 0.000830, ‖Δτ‖ = 0.025631, ‖ΔL‖ = 0.000042, status = LOCALLY_SOLVED, time(s) = 0.043087
run = 1, ‖Δβ‖ = 0.683108, ‖Δτ‖ = 0.773919, ‖ΔL‖ = 0.042397, status = LOCALLY_SOLVED, time(s) = 0.044666
run = 2, ‖Δβ‖ = 0.008159, ‖Δτ‖ = 0.069178, ‖ΔL‖ = 0.003574, status = LOCALLY_SOLVED, time(s) = 0.047489






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(CSV.read(&quot;gxe_snp.pval.txt&quot;, DataFrame), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m   chr [0m|[1m     pos [0m|[1m      snpid [0m|[1m       maf [0m|[1m    hwepval [0m|[1m snpeffectnullbeta [0m|[1m snpeffectnulltau [0m|[1m betapval [0m|[1m  taupval [0m|
|[90m Int64 [0m|[90m   Int64 [0m|[90m   String15 [0m|[90m   Float64 [0m|[90m    Float64 [0m|[90m           Float64 [0m|[90m          Float64 [0m|[90m  Float64 [0m|[90m  Float64 [0m|
|-------|---------|------------|-----------|------------|-------------------|------------------|----------|----------|
|     1 |  554484 | rs10458597 |       0.0 |        1.0 |               0.0 |              0.0 |      1.0 |      1.0 |
|     1 |  758311 | rs12562034 | 0.0776398 |   0.409876 |         -0.230593 |        0.0102355 | 0.448376 | 0.149797 |
|     1 |  967643 |  rs2710875 |  0.324074 | 4.07625e-7 |          0.639871 |        -0.338815 | 0.293876 | 0.119916 |
|     1 | 1168108 | rs11260566 |  0.191589 |   0.128568 |          0.618855 |        -0.146819 | 0.590189 | 0.262791 |
|     1 | 1375074 |  rs1312568 |  0.441358 | 2.5376e-19 |         -0.456003 |         0.492582 | 0.574737 | 0.250501 |</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;gxe_snp.pval.txt&quot;, force=true)</code></pre><h2 id="SNP-set-testing"><a class="docs-heading-anchor" href="#SNP-set-testing">SNP-set testing</a><a id="SNP-set-testing-1"></a><a class="docs-heading-anchor-permalink" href="#SNP-set-testing" title="Permalink"></a></h2><p>In many applications, we want to test a SNP-set. The function with keyword <code>analysistype = &quot;snpset&quot;</code> can be used to do this. To specify the type of snpset test, use the <code>snpset</code> keyword argument. </p><p>The snpset can be specified as either:</p><ul><li>a window (test every X snps) =&gt; <code>snpset = X</code></li><li>an annotated file.  This requires <code>snpset = filename</code> where filename is an input file, with no header and two columns separated by a space. The first column must contain the snpset ID and the second column must contain the snpid&#39;s (rsid) identical to the bimfile, or in the case of a BGEN format, the order the data will be read (see BGEN above for more details).</li><li>a joint test on only a specific set of snps. <code>snpset = AbstractVector</code> where the vector specifies the snps you want to perform one joint snpset test for. The vector can either be a vector of integers where the elements are the indicies of the SNPs to test, a vector of booleans, where true represents that you want to select that SNP index, or a range indicating the indicies of the SNPs to test. </li></ul><p>SNPset testing is currently only implemented for the score test (not Wald). </p><p>In the following example, we perform a SNP-set test on the 50th to 55th snps. </p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;snpset.pval.txt&quot;, 
        analysistype = &quot;snpset&quot;,
        snpset = 50:55)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043069
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040745






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">run(`head snpset.pval.txt`);</code></pre><pre><code class="nohighlight hljs">The pvalue of snps indexed at 50:55 is betapval: 2.315617121196509e-5, taupval: 3.972993706371914e-10</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;trajgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on the annotated SNP-set file.</p><pre><code class="language-julia hljs">run(`head ../data/hapmap_snpsetfile.txt`);</code></pre><pre><code class="nohighlight hljs">gene1 rs10458597
gene1 rs12562034
gene1 rs2710875
gene1 rs11260566
gene1 rs1312568
gene1 rs35154105
gene1 rs16824508
gene1 rs2678939
gene1 rs7553178
gene1 rs13376356</code></pre><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;snpset.pval.txt&quot;, 
        analysistype = &quot;snpset&quot;,
        snpset = datadir * &quot;/hapmap_snpsetfile.txt&quot;)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044896
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040974






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;snpset.pval.txt&quot;, DataFrame; delim=&quot;\t&quot;), 5), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m snpsetid [0m|[1m nsnps [0m|[1m  betapval [0m|[1m   taupval [0m|
|[90m  String7 [0m|[90m Int64 [0m|[90m   Float64 [0m|[90m   Float64 [0m|
|----------|-------|-----------|-----------|
|    gene1 |    93 |  0.111864 |  0.011595 |
|    gene2 |    93 | 0.0249929 | 0.0648265 |
|    gene3 |    93 |  0.131741 | 0.0298884 |
|    gene4 |    92 |  0.010327 | 0.0584591 |
|    gene5 |    93 | 0.0303905 | 0.0924954 |</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;trajgwas.null.txt&quot;, force=true)</code></pre><p>In the following example we run a SNP-set test on every 15 SNPs.</p><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = &quot;snpset.pval.txt&quot;, 
        analysistype = &quot;snpset&quot;,
        snpset = 15)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043705
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.039385






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs">pretty_table(first(CSV.read(&quot;snpset.pval.txt&quot;, DataFrame), 5), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m startchr [0m|[1m startpos [0m|[1m startsnpid [0m|[1m endchr [0m|[1m   endpos [0m|[1m   endsnpid [0m|[1m    betapval [0m|[1m     taupval [0m|
|[90m    Int64 [0m|[90m    Int64 [0m|[90m   String15 [0m|[90m  Int64 [0m|[90m    Int64 [0m|[90m   String15 [0m|[90m     Float64 [0m|[90m     Float64 [0m|
|----------|----------|------------|--------|----------|------------|-------------|-------------|
|        1 |   554484 | rs10458597 |      1 |  3431124 | rs12093117 |  3.83046e-6 | 2.24303e-10 |
|        1 |  3633945 | rs10910017 |      1 |  6514524 |   rs932112 | 0.000127922 |  7.94764e-6 |
|        1 |  6715827 |   rs441515 |      1 |  9534606 |  rs4926480 |  8.10353e-5 |  6.11848e-7 |
|        1 |  9737551 | rs12047054 |      1 | 12559747 |  rs4845907 | 0.000425108 |  1.15311e-8 |
|        1 | 12760427 |   rs848577 |      1 | 16021797 |  rs6679870 |  2.75832e-5 |  0.00028553 |</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;snpset.pval.txt&quot;, force=true)
rm(&quot;trajgwas.null.txt&quot;, force=true)</code></pre><h2 id="Matching-Indicies"><a class="docs-heading-anchor" href="#Matching-Indicies">Matching Indicies</a><a id="Matching-Indicies-1"></a><a class="docs-heading-anchor-permalink" href="#Matching-Indicies" title="Permalink"></a></h2><p>In some cases, there are only a subset of individuals with both genetic data and covariate information available. The null model must be fit on a subset of the individuals with the genetic data. The rows can be specified with the argument <code>covrowinds</code> if you pass in a covariate file. The genetic indicies can be specified with the <code>geneticrowinds</code> argument. </p><p>For simplicity, we have implemented a function <code>matchindices(meanformula, reformula, wsvarformula, idvar, df, geneticsamples)</code> which can be used to do this. Input the mean, random effects, and within-subject variance formulas, the grouping (id) variable, the dataframe (or table), and a vector of the IDs in the order of the genetic file and it returns <code>covrowmask, geneticrowmask</code> for matching indicies in a covariate file and geneticfile.</p><p>Note: the <code>idvar</code> in the dataframe and the <code>geneticsamples</code> vector must have the same element type. To convert a vector of integers to strings <code>string.(intvector)</code> can be used. To go from a vector of strings to integers you can use&quot;  <code>map(x -&gt; parse(Int, x), stringvectoparse)</code>.</p><pre><code class="language-julia hljs">famfileids = CSV.read(datadir * &quot;hapmap3.fam&quot;, DataFrame, header = false)[!, 1] # famfile contains the sample IDs for PLINK files</code></pre><pre><code class="nohighlight hljs">324-element Vector{String7}:
 &quot;A1&quot;
 &quot;2&quot;
 &quot;3&quot;
 &quot;4&quot;
 &quot;5&quot;
 &quot;6&quot;
 &quot;7&quot;
 &quot;8&quot;
 &quot;9&quot;
 &quot;10&quot;
 &quot;11&quot;
 &quot;12&quot;
 &quot;13&quot;
 ⋮
 &quot;313&quot;
 &quot;314&quot;
 &quot;315&quot;
 &quot;316&quot;
 &quot;317&quot;
 &quot;318&quot;
 &quot;319&quot;
 &quot;320&quot;
 &quot;321&quot;
 &quot;322&quot;
 &quot;323&quot;
 &quot;Z324&quot;</code></pre><pre><code class="language-julia hljs">covdf = CSV.read(datadir * &quot;trajgwas_plinkex.csv&quot;, DataFrame)
pretty_table(first(covdf, 11), tf = tf_markdown)</code></pre><pre><code class="nohighlight hljs">|[1m     sex [0m|[1m  onMeds [0m|[1m    snp1 [0m|[1m    snp2 [0m|[1m    snp3 [0m|[1m    snp4 [0m|[1m       y [0m|[1m      id [0m|
|[90m Float64 [0m|[90m Float64 [0m|[90m Float64 [0m|[90m Float64 [0m|[90m Float64 [0m|[90m Float64 [0m|[90m Float64 [0m|[90m String7 [0m|
|---------|---------|---------|---------|---------|---------|---------|---------|
|     0.0 |     1.0 |     0.0 |     1.0 |     2.0 |     0.0 | 12.2667 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 10.2681 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 |  12.166 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 11.8797 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 12.8127 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 9.98766 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 12.1408 |      A1 |
|     0.0 |     1.0 |     0.0 |     1.0 |     2.0 |     0.0 | 13.2058 |      A1 |
|     0.0 |     0.0 |     0.0 |     1.0 |     2.0 |     0.0 | 11.3631 |      A1 |
|     0.0 |     1.0 |     0.0 |     1.0 |     2.0 |     0.0 | 15.2511 |      A1 |
|     0.0 |     0.0 |     0.0 |     0.0 |     2.0 |     2.0 |  12.746 |       2 |</code></pre><pre><code class="language-julia hljs">covrowmask, geneticrowmask = matchindices(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id, covdf, famfileids);</code></pre><pre><code class="language-julia hljs">trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        datadir * &quot;hapmap3&quot;,
        pvalfile = pvalpath,
        nullfile = nullpath,
        covrowinds = covrowmask,
        geneticrowinds = geneticrowmask)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.043990
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.040579






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
rm(&quot;trajgwas.pval.txt&quot;, force=true)</code></pre><h2 id="Plotting-Results"><a class="docs-heading-anchor" href="#Plotting-Results">Plotting Results</a><a id="Plotting-Results-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-Results" title="Permalink"></a></h2><p>To plot the GWAS results, we recommend using the <a href="https://openmendel.github.io/MendelPlots.jl/latest/">MendelPlots.jl package</a>.</p><h2 id="Multiple-Plink-file-sets"><a class="docs-heading-anchor" href="#Multiple-Plink-file-sets">Multiple Plink file sets</a><a id="Multiple-Plink-file-sets-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-Plink-file-sets" title="Permalink"></a></h2><p>In large scale studies, genotypes data are split into multiple Plink files, e.g., by chromosome. Then GWAS analysis can be done in parallel. This can be achieved by two steps.</p><p>Let&#39;s first create demo data by splitting hapmap3 according to chromosome:</p><pre><code class="language-julia hljs"># split example hapmap3 data according to chromosome
SnpArrays.split_plink(datadir * &quot;hapmap3&quot;, :chromosome; prefix=datadir * &quot;hapmap3.chr.&quot;)
readdir(glob&quot;hapmap3.chr.*&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">75-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.1.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.1.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.1.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.10.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.10.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.10.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.11.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.11.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.11.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.12.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.12.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.12.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.13.bed&quot;
 ⋮
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.6.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.6.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.6.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.7.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.7.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.7.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.8.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.8.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.8.fam&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.9.bed&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.9.bim&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.9.fam&quot;</code></pre><p>Step 1: Fit the null model. Setting third argument <code>geneticfile</code> to <code>nothing</code> instructs <code>trajgwas</code> function to fit the null model only.</p><pre><code class="language-julia hljs">nm = trajgwas(@formula(y ~ 1 + sex + onMeds),
        @formula(y ~ 1),
        @formula(y ~ 1 + sex + onMeds),
        :id,
        datadir * &quot;trajgwas_plinkex.csv&quot;,
        nothing)</code></pre><pre><code class="nohighlight hljs">run = 1, ‖Δβ‖ = 0.037090, ‖Δτ‖ = 0.136339, ‖ΔL‖ = 0.005441, status = LOCALLY_SOLVED, time(s) = 0.044113
run = 2, ‖Δβ‖ = 0.000913, ‖Δτ‖ = 0.019810, ‖ΔL‖ = 0.001582, status = LOCALLY_SOLVED, time(s) = 0.038771






Within-subject variance estimation by robust regression (WiSER)

Mean Formula:
y ~ 1 + sex + onMeds
Random Effects Formula:
y ~ 1
Within-Subject Variance Formula:
y ~ 1 + sex + onMeds

Number of individuals/clusters: 324
Total observations: 3240

Fixed-effects parameters:
────────────────────────────────────────────────────────
                  Estimate  Std. Error       Z  Pr(&gt;|Z|)
────────────────────────────────────────────────────────
β1: (Intercept)  13.2282     0.146459    90.32    &lt;1e-99
β2: sex          -3.29295    0.2101     -15.67    &lt;1e-54
β3: onMeds        0.459585   0.0596002    7.71    &lt;1e-13
τ1: (Intercept)   0.792508   0.0850728    9.32    &lt;1e-19
τ2: sex          -0.2865     0.0970732   -2.95    0.0032
τ3: onMeds        0.422303   0.063825     6.62    &lt;1e-10
────────────────────────────────────────────────────────
Random effects covariance matrix Σγ:
 &quot;γ1: (Intercept)&quot;  3.32057</code></pre><p>Step 2: GWAS for each chromosome.</p><pre><code class="language-julia hljs"># this part can be submitted as separate jobs
for chr in 1:23
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    pvalfile = plinkfile * &quot;.pval.txt&quot; 
    trajgwas(nm, plinkfile, pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia hljs"># show the result files
readdir(glob&quot;*.pval.txt&quot;, datadir)</code></pre><pre><code class="nohighlight hljs">23-element Vector{String}:
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.1.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.10.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.11.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.12.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.13.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.14.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.15.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.16.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.17.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.18.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.19.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.2.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.20.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.21.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.22.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.23.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.3.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.4.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.5.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.6.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.7.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.8.pval.txt&quot;
 &quot;/Users/xyz/.julia/dev/TrajGWAS/data/hapmap3.chr.9.pval.txt&quot;</code></pre><p>In the rare situations where the multiple sets of Plink files lack the <code>fam</code> file or the corresponding bed and bim files have different filenames, users can explicitly supply bed filename, bim file name, and number of individuals. Replace Step 2 by </p><p>Step 2&#39;: GWAS for each chromosome.</p><pre><code class="language-julia hljs"># this part can be submitted as separate jobs
for chr in 1:23
    bedfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bed&quot;
    bimfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.bim&quot;
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    trajgwas(nm, bedfile, bimfile, 324; pvalfile=pvalfile)
end</code></pre><pre><code class="language-julia hljs"># clean up
rm(&quot;trajgwas.null.txt&quot;, force=true)
isfile(datadir * &quot;fittednullmodel.jld2&quot;) &amp;&amp; rm(datadir * &quot;fittednullmodel.jld2&quot;)
for chr in 1:23
    pvalfile = datadir * &quot;hapmap3.chr.&quot; * string(chr) * &quot;.pval.txt&quot;
    rm(pvalfile, force=true)
end
for chr in 1:26
    plinkfile = datadir * &quot;hapmap3.chr.&quot; * string(chr)
    rm(plinkfile * &quot;.bed&quot;, force=true)
    rm(plinkfile * &quot;.fam&quot;, force=true)
    rm(plinkfile * &quot;.bim&quot;, force=true)
end</code></pre><h2 id="Multiple-file-sets-on-cluster"><a class="docs-heading-anchor" href="#Multiple-file-sets-on-cluster">Multiple file sets on cluster</a><a id="Multiple-file-sets-on-cluster-1"></a><a class="docs-heading-anchor-permalink" href="#Multiple-file-sets-on-cluster" title="Permalink"></a></h2><p>For running the score tests on a cluster, it would be desirable to fit a null model on a single machine first, and save the model as a serialized Julia object (<code>.jls</code>). For example:</p><pre><code class="language-julia hljs">using DataFrames, CSV
using Statistics
using TrajGWAS
using Ipopt, WiSER
using LinearAlgebra
using BGEN
# fit the null model
BLAS.set_num_threads(1)
solver = Ipopt.Optimizer()
solver_config = Dict(&quot;print_level&quot; =&gt; 1, 
    &quot;watchdog_shortened_iter_trigger&quot; =&gt; 5,
    &quot;max_iter&quot; =&gt; 120)

bp_data = CSV.read(&quot;bp.csv&quot;, DataFrame)
@time nm = trajgwas(@formula(dbp ~ 1 + SEX + age + age_sq +
        PC1 + PC2 + PC3 + PC4 + PC5 + bmi),
    @formula(dbp ~ 1 + age),
    @formula(dbp ~ 1 + SEX + age + age_sq +
        PC1 + PC2 + PC3 + PC4 + PC5 +
        bmi),
    :FID, # subject ID
    bp_data,
    nothing;
    nullfile=&quot;dbp.null.txt&quot;,
    solver=solver,
    solver_config=solver_config,
    init = nothing, # may change to `x -&gt; WiSER.init_ls!(x; gniters=0)`
    runs=10
)

println(nm)
using Serialization
open(&quot;null.model.jls&quot;, &quot;w&quot;) do io
    Serialization.serialize(io, nm)
end</code></pre><p>Then, the fitted null model would be used for the score test. It could be desirable to configure each job to run on a slice of SNPs on a BGEN file for higher throughput. The julia script for score test (<code>scoretest_bp.jl</code>) for a slice of BGEN file would look like:</p><pre><code class="language-julia hljs">using DataFrames, CSV
using Statistics
using TrajGWAS
using WiSER
using LinearAlgebra
using BGEN
# fit the null model
BLAS.set_num_threads(1)


using Serialization

bgendir = ARGS[1] 
chr = ARGS[2] # 1 to 22
fitted_null = ARGS[3] # &quot;null.model.jls&quot;
pvalfile = ARGS[4] # &quot;sbp.test.diabetics.chr$(chr).txt&quot;
chunkidx = parse(Int, ARGS[5])
nchunks  = parse(Int, ARGS[6])

nm = open(deserialize, fitted_null)
genetic_iids_subsample = nm.ids

bgenfilename = bgendir * &quot;/Chr$(chr)&quot; # to analyze, for example, /path/to/bgen/Chr5.bgen
samplefilename = bgendir * &quot;/Samples.sample&quot; # .sample file compatible with BGEN
mfifilename = bgendir * &quot;/mfi_chr$(chr).txt&quot; # &quot;MFI&quot; file, an external file with MAF + info score	
ukb_data = Bgen(bgenfilename * &quot;.bgen&quot;; sample_path = samplefilename)
genetic_iids = map(x -&gt; parse(Int, split(x, &quot; &quot;)[1]), samples(ukb_data))

order_dict = Dict{Int, Int}()
for (i, iid) in enumerate(genetic_iids)
    order_dict[iid] = i
end

sample_indicator = falses(length(genetic_iids))
for v in genetic_iids_subsample
    sample_indicator[order_dict[v]] = true # extract only the samples being used for the analysis
end

# GWAS for each chromosome

## pre-filtering SNPs not passing the criteria (MAF &gt; 0.002, info score &gt; 0.3)
min_maf = 0.002
min_info_score = 0.3
min_hwe_pval = 1e-10
mfi = CSV.read(mfifilename, DataFrame; header=false)
mfi.Column8 = map(x -&gt; x == &quot;NA&quot; ? NaN : parse(Float64, x), mfi.Column8) # Column8: info score
snpmask = (mfi.Column6 .&gt; min_maf) .&amp; (mfi.Column8 .&gt; 0.3) # Column6: MAF

## compute range to run the analysis
chunksize = n_variants(ukb_data) ÷ nchunks + (n_variants(ukb_data) % nchunks &gt; 0 ? 1 : 0)
startidx = chunksize * (chunkidx - 1) + 1
endidx = min(chunksize * chunkidx, n_variants(ukb_data))
snpmask = snpmask[startidx:endidx]

println(&quot;running for variants $startidx to $endidx&quot;)

# rearrange data in nm so that it matches bgen data
nullinds = indexin(genetic_iids[sample_indicator], nm.ids)
nm.obswts .= isempty(nm.obswts) ? nm.obswts : nm.obswts[nullinds]
nm.ids .= nm.ids[nullinds]
nm.nis .= nm.nis[nullinds]
nm.data .= nm.data[nullinds]
@assert genetic_iids[sample_indicator] == nm.ids &quot;there is some issue -- sampleids not matching&quot;
    
trajgwas(nm, bgenfilename * &quot;.bgen&quot;, count(sample_indicator);
    samplepath=samplefilename,
    pvalfile=pvalfile,
    snpinds=snpmask,
    min_hwe_pval = min_hwe_pval,
    bgenrowinds = sample_indicator,
    startidx = startidx,
    endidx = endidx,
    usespa=true)
</code></pre><p>Note that an index file (<code>.bgen.bgi</code>) is required for this slicing of BGEN file. See <a href="https://enkre.net/cgi-bin/code/bgen/doc/trunk/doc/wiki/bgenix.md">this link</a> to see how to create one. </p><ul><li>Command-line arguments<ul><li>Argument 1: directory for the BGEN files. BGEN files (.bgen), BGEN index files (.bgen.bgi), and MFI files (.txt) should be included there.</li><li>Argument 2: chromosome</li><li>Argument 3: fitted null model (.jls)</li><li>Argument 4: path for the result p-value file</li><li>Argument 5: chunk index (1-based)</li><li>Argument 6: number of chunks</li></ul></li></ul><p>The code above runs the analysis on <code>ARGS[5]</code>-th slice out of <code>ARGS[6]</code> slices of chromosome <code>ARGS[2]</code>.</p><p>Then, the following script could be used for a cluster managed by Sun Grid Engine: (<code>sbp_diabetes.sh</code>)</p><pre><code class="language-bash hljs">#!/bin/bash
#$ -cwd
#$ -o joblog.$JOB_ID.$TASK_ID
#$ -j y
#$ -pe shared 2
#$ -l h_rt=8:00:00,h_data=8G,arch=intel*
# Email address to notify
##$ -M $USER@mail
# Notify when
#$ -m a
#  Job array indexes
#$ -t 1-352:1

NCHUNKS=16
CHUNKIDX=$(( (${SGE_TASK_ID} - 1) % ${NCHUNKS} + 1 ))
CHR=$(( (${SGE_TASK_ID} - 1) / ${NCHUNKS} + 1))

PROJECTDIR=/user/dir/jobscripts
BGENDIR=/user/dir/imputed
FITTED_NULL=/user/dir/null.model.jls
PVALFILE=/user/dir/pvalfiles/sbp.test.diabetes.chr${CHR}.${CHUNKIDX}of${NCHUNKS}.txt

module load julia
time julia --project=${PROJECTDIR} ${PROJECTDIR}/scoretest_bp.jl ${BGENDIR} ${CHR} ${FITTED_NULL} ${PVALFILE} ${CHUNKIDX} ${NCHUNKS}</code></pre><p>and this could be submitted using the <code>qsub</code> command.</p><h2 id="Troubleshooting"><a class="docs-heading-anchor" href="#Troubleshooting">Troubleshooting</a><a id="Troubleshooting-1"></a><a class="docs-heading-anchor-permalink" href="#Troubleshooting" title="Permalink"></a></h2><p>If there are issues you&#39;re encountering with running TrajGWAS, the following are possible remedies.</p><ul><li>Null Model<ul><li>Issues with null model convergence may be solved by choosing alternate starting values for parameters, using a different solver, transforming variables, and increasing the number of runs (WiSER runs). These are detailed in the <a href="https://github.com/OpenMendel/WiSER.jl/blob/master/docs/src/model_fitting.md#tips-for-improving-estimation">WiSER documentation here</a>. <code>init = x -&gt; WiSER.init_ls!(x; gniters=0)</code> works fine in general. </li></ul></li><li>GWAS Results<ul><li>If you use the score test instead of the SPA-score test (SPA is default for single-SNP analyses), then there can be inflation in type I error and decreased power when (a) the sample size is small, (b) the number of repeated measures is low, or (c) the variants analyzed are rare with low minor allele frequencies. In these cases, the score test is not optimal and it is suggested to use the SPA version (<code>usespa=true</code>). SPA is only implemented for single-SNP analyses. These issues can occur in both Wald and score tests. </li></ul></li></ul><p>If you notice any problems with your output or results, <a href="https://github.com/OpenMendel/TrajGWAS.jl/issues">file an issue</a>. </p></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Friday 4 November 2022 21:41">Friday 4 November 2022</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
